homeassistant/components/climate/generic_thermostat.py

```
@@ -302,65 +390,143 @@ def _async_update_temp(self, state):
+                current_state = STATE_ON
+            else:
+                current_state = STATE_OFF
+            long_enough_heating = condition.state(
                 self.hass, self.heater_entity_id, current_state,
                 self.min_cycle_duration)
-            if not long_enough:
-                return
-
-        if self.ac_mode:
-            is_cooling = self._is_device_active
-            if is_cooling:
-                too_cold = self._target_temp - self._cur_temp >= \
-                    self._cold_tolerance
-                if too_cold:
-                    _LOGGER.info('Turning off AC %s', self.heater_entity_id)
-                    self._heater_turn_off()
+            if self.current_operation == STATE_COOL:
+                if not long_enough_cooling:
+                    return
+            if self.current_operation == STATE_HEAT:
+                if not long_enough_heating:
+                    return
+            if self.current_operation == STATE_AUTO:
+                if not long_enough_cooling or not long_enough_heating:
+                    return
+
+        if self.current_operation == STATE_COOL:
+            if self._dual_temp:
+                target_temp = self._target_temp_high
             else:
-                too_hot = self._cur_temp - self._target_temp >= \
-                    self._hot_tolerance
-                if too_hot:
-                    _LOGGER.info('Turning on AC %s', self.heater_entity_id)
-                    self._heater_turn_on()
-        else:
-            is_heating = self._is_device_active
-            if is_heating:
-                too_hot = self._cur_temp - self._target_temp >= \
-                    self._hot_tolerance
-                if too_hot:
-                    _LOGGER.info('Turning off heater %s',
-                                 self.heater_entity_id)
-                    self._heater_turn_off()
+                target_temp = self._target_temp
+            if self.ac_entity_id:
```

`STATE_COOL` should never be settable as an operation mode if we don't have an `ac_entity_id`. Thus, you can get rid of this `if`. If you want to make super-sure, you could check, throw an error and exit the method here if `ac_entity_id` is not set. However, the level of nesting is getting a bit out of hand in this method, I think. ;)

---

I've removed state, and just used entity check. This way I could reuse logic for all states (AUTO/COOL/HEAT) because OFF state returns earlier.